File: 0712-minimum-ascii-delete-sum-for-two-strings/0712-minimum-ascii-delete-sum-for-two-strings.cpp
class Solution {
public:
    int minimumDeleteSum(string s1, string s2) {
        vector<vector<int>> dp(s1.length(), vector<int>(s2.length(), -1));
        return solve(s1, s2, dp, 0, 0);
    }
private:
    int solve(string& s1, string& s2, vector<vector<int>>& dp, int i, int j)
    {
        if(i == s1.length() || j == s2.length())
        {
            int rem = 0;
            while(i < s1.length())
            {
                rem += s1[i];
                i++;
            }
            while(j < s2.length())
            {
                rem += s2[j];
                j++;
            }
            return rem;
        }
        if(dp[i][j] == -1)
        {
            dp[i][j] = INT_MAX;
            if(s1[i] == s2[j])
            {
                dp[i][j] = min(dp[i][j], solve(s1, s2, dp, i + 1, j + 1));
            }
            else
            {
                dp[i][j] = min(dp[i][j], solve(s1, s2, dp, i + 1, j) + s1[i]);
                dp[i][j] = min(dp[i][j], solve(s1, s2, dp, i, j + 1) + s2[j]);
            }
        }
        return dp[i][j];
    }
};

File: 0712-minimum-ascii-delete-sum-for-two-strings/0712-minimum-ascii-delete-sum-for-two-strings.py
def solve(s1: str, s2: str, dp: List[List[int]], i: int, j: int) -> int:
    if i == len(s1) or j == len(s2):
        rem = 0
        while i < len(s1):
            rem += ord(s1[i])
            i += 1
        while j < len(s2):
            rem += ord(s2[j])
            j += 1
        return rem
    if dp[i][j] == -1:
        dp[i][j] = float('inf')
        if s1[i] == s2[j]:
            dp[i][j] = min(dp[i][j], solve(s1, s2, dp, i + 1, j + 1))
        else:
            dp[i][j] = min(dp[i][j], solve(s1, s2, dp, i + 1, j) + ord(s1[i]))
            dp[i][j] = min(dp[i][j], solve(s1, s2, dp, i, j + 1) + ord(s2[j]))
    return dp[i][j]
class Solution:
    def minimumDeleteSum(self, s1: str, s2: str) -> int:
        dp = [[-1 for _ in range(len(s2))] for _ in range(len(s1))]
        return solve(s1, s2, dp, 0, 0)
        